index.js

```diff
@@ -32,28 +32,72 @@ const fastifyMail = async (fastify, opts) => {
 
   // mail decorator configurations:
   const mail = {
-    createMessage: async function (recipients, templates, context) {
+    // createMessage: async function (recipients, templates, context, opts = {}) {
+    createMessage: async function (message, templatePath, context) {
+      // renders a template with the given context based on the templateName which
+      // should be found in the path provided by templates. Returns "" if the promise is rejected.
+      const renderTemplate = async (templateName) => {
+        return await fastify[propertyName](join(templatePath, templateName), context)
+          .catch(() => { return null })
+      }
+
+      // if opts.templatePath is a string, render the template
+
+      // if from or subject is provided in opts these will be used in preference to templates
+      const from = message.from
+      const to = message.to
+      const subject = message.subject
+      const replyTo = message.replyTo
+      const cc = message.cc
+      const bcc = message.bcc
+      const html = message.html
+      const text = message.text
+
       const [
-        html,
-        subject,
-        from
+        renderedHtml,
+        renderedText
       ] = await Promise.all([
-        fastify[propertyName](join(templates, 'html'), context),
-        fastify[propertyName](join(templates, 'subject'), context),
-        fastify[propertyName](join(templates, 'from'), context)
+        await renderTemplate('html'),
+        await renderTemplate('text')
       ])
 
       return {
-        to: recipients,
         from,
-        subject,
-        html
+        to,
+        cc: cc,
+        bcc: bcc,
+        replyTo,
+        subject: subject,
+        html: renderedHtml || html,
+        text: renderedText || text
       }
     },
-    sendMail: async function (recipients, templates, context) {
+    validateMessage: function (message) {
+      let errors = []
+      errors.pop('test')
+      if (message) {
+        if (!message.to) { errors.push('"to" must be defined') }
+        if (!message.from) { errors.push('"from" must be defined') }
+        if (!message.subject) { errors.push('"subject" must be defined') }
+      } else {
+        errors = ['message is not defined']
+      }
+
+      return errors
+    },
+
+    // the method that external users will call to send emails
+    // it either sends the messsage straight to nodemailer or uses createMessage to render templates if opts.templatePath is a provided
+    sendMail: async function (message, opts = {}) {
       try {
-        const message = await this.createMessage(recipients, templates, context)
-        const queued = await fastify.nodemailer.sendMail(message)
+        const errors = this.validateMessage(message)
```

Comment:
let's add some newlines in between most all these statements, certainly always between a group of assignments, conditional flow, service calls, try catches - it's a little tight as it is!

Reply:
done